src/iterative_ensemble_smoother/esmda.py

```diff
@@ -0,0 +1,255 @@
+        self.observations = observations
+        self.iteration = 0
+        self.rng = np.random.default_rng(seed)
+        self.inversion = inversion
+
+        # Alpha can either be an integer (num iterations) or a list of weights
+        if isinstance(alpha, np.ndarray) and alpha.ndim == 1:
+            self.alpha = normalize_alpha(alpha)
+        elif isinstance(alpha, numbers.Integral):
+            self.alpha = normalize_alpha(np.ones(alpha))
+            assert np.allclose(self.alpha, normalize_alpha(self.alpha))
+        else:
+            raise TypeError("Alpha must be integer or 1D array.")
+
+        # Only compute the covariance factorization once
+        # If it's a full matrix, we gain speedup by only computing cholesky once
+        # If it's a diagonal, we gain speedup by never having to compute cholesky
+        num_outputs = C_D.shape[0]
+
+        if isinstance(C_D, np.ndarray) and C_D.ndim == 2:
+            self.C_D_L = sp.linalg.cholesky(C_D, lower=False)
+        elif isinstance(C_D, np.ndarray) and C_D.ndim == 1:
+            assert len(C_D) == num_outputs
+            self.C_D_L = np.sqrt(C_D)
+        else:
+            raise TypeError("Argument `C_D` must be 1D or 2D array")
+
+        self.C_D = C_D
+        assert isinstance(self.C_D, np.ndarray) and self.C_D.ndim in (1, 2)
+
+    def num_assimilations(self) -> int:
+        return len(self.alpha)
+
+    def assimilate(
+        self,
+        X: npt.NDArray[np.double],
+        Y: npt.NDArray[np.double],
+        ensemble_mask: Optional[npt.NDArray[np.bool_]] = None,
+        overwrite: bool = False,
+        truncation: float = 1.0,
+    ) -> npt.NDArray[np.double]:
+        """Assimilate data and return an updated ensemble X_posterior.
+
+        Parameters
+        ----------
+        X : np.ndarray
+            2D array of shape (num_inputs, num_ensemble_members).
+        Y : np.ndarray
+            2D array of shape (num_ouputs, num_ensemble_members).
+        ensemble_mask : np.ndarray
+            1D boolean array of length `num_ensemble_members`, describing which
+            ensemble members are active. Inactive realizations are ignored.
+            Defaults to all active.
+        overwrite : bool
+            If True, then arguments X and Y may be overwritten.
+            If False, then the method will not permute inputs in any way.
+        truncation : float
+            How large a fraction of the singular values to keep in the inversion
+            routine. Must be a float in the range (0, 1]. A lower number means
+            a more approximate answer and a slightly faster computation.
+
+        Returns
+        -------
+        X_posterior : np.ndarray
+            2D array of shape (num_inputs, num_ensemble_members).
+
+        """
+        if self.iteration >= self.num_assimilations():
+            raise Exception("No more assimilation steps to run.")
+
+        # Verify shapes
+        num_inputs, num_ensemble = X.shape
```

Comment:
`num_inputs` is not used

Reply:
nice find. i ran [vulture](https://pypi.org/project/vulture/) and removed this, as well as some other, unused variables